File: 백준/Silver/24444. 알고리즘 수업 － 너비 우선 탐색 1/알고리즘 수업 － 너비 우선 탐색 1.py
import sys
input = sys.stdin.readline

N, M, R = map(int, input().split())
arr = [[] for _ in range(N+1)]

for _ in range(M):
    a, b = map(int, input().split())
    arr[a].append(b)
    arr[b].append(a)

def bfs():
    q = [R]
    visited = [0] * (N+1)
    cnt = 1
    while q:
        x = q.pop(0)

        if visited[x] == 0:
            visited[x] = cnt

        arr[x].sort()
        for i in arr[x]:
            if visited[i] == 0:
                visited[i] = cnt + 1
                cnt += 1

                q.append(i)

    return visited

x = bfs()
for i in range(1, len(x)):
    print(x[i])

File: 백준/Silver/24444. 알고리즘 수업 － 너비 우선 탐색 1/알고리즘 수업 － 너비 우선 탐색 1.js
let fs = require('fs')
let input = fs.readFileSync('/dev/stdin').toString().trim().split('\n')
// let input = fs.readFileSync('./EX.txt').toString().trim().split('\n')

const [N, M, R] = input[0].split(" ").map(x => parseInt(x))
let arr = Array.from({length : N+1}, () => [])

for (let i = 1; i <= M; i++) {
    const [a, b] = input[i].split(" ").map(x => parseInt(x))
    arr[a].push(b)
    arr[b].push(a)
}

function bfs(x) {
    
    let visited = Array(N+1).fill(0)

    let q = [x]
    let cnt = 1
    while (q.length) {
        let y = q.shift()


        if (visited[y] === 0) {
            visited[y] = cnt
        }
        arr[y].sort((a, b) => a - b)
        // console.log(arr[y])
        arr[y].forEach(x => {
            if (visited[x] === 0) {
                visited[x] = cnt + 1
                q.push(x)
                cnt += 1
            }
        })
    }
    return visited
}

result = bfs(R)
ans = ''
for (let i = 1; i < result.length; i++) {
    ans += result[i] + '\n'
}
console.log(ans)